carddav/server.go

```diff
@@ -406,6 +454,14 @@ func (b *backend) Proppatch(r *http.Request, update *internal.Propertyupdate) (*
 }
 
 func (b *backend) Put(r *http.Request) (*internal.Href, error) {
+	homeSetPath, err := b.Backend.AddressbookHomeSetPath(r.Context())
+	if err != nil {
+		return nil, err
+	}
+	if !strings.HasPrefix(r.URL.Path, homeSetPath) {
+		return nil, &internal.HTTPError{Code: http.StatusMethodNotAllowed}
+	}
```

Comment:
I'm not sure it's a good idea to add this logic. Servers may want to allow users to fetch/edit calendars which aren't under the home set. Additionally, `strings.HasPrefix` is not appropriate here -- the check passes with the path `/abcdef` and the home set `/abc`.

Reply:
Ok, so a few things :slightly_smiling_face: 

> Servers may want to allow users to fetch/edit calendars which aren't under the home set

I agree that this won't work like this, but as things are right now, this is outright impossible anyways (due to the home set being `/`). I think part of why this PR turned a bit tedious might be that I cannot quite seem to gauge your expectations. I initially had the impression you'd actually prefer to only do the bare minimum required to get the two servers working side-by-side, which would have just been allowing a home set prefix. I guess it is on me that I brought the user principal URL into this. 

However, with this comment, I get the impression that now we're talking about laying the foundations for all kinds of things that have been previously impossible, such as accessing other people's calendars. This is fine, I just need to understand where we want to go.

Through #70 , I also realized that the backends should behave more like a HTTP backend (as opposed to e.g. map lookup or such). As such, my suggestion would be:

* we remove that check entirely
* as discussed in #70, we expose some appropriate internal errors (not found, access denied, etc)
* we document that the backend can have absolutely zero expectations w/ regards to the path it is asked about and has to check it itself and return the appropriate errors

WDYT?

PS: on a side note, I am quite disappointed that Go doesn't have a proper `path.HasPrefix()`, which is why I initially wrote those custom functions. But the check really would have never been good enough anyways, and the backend will have to perform them. So just removing it from here really makes sense...

Reply:
> However, with this comment, I get the impression that now we're talking about laying the foundations for all kinds of things that have been previously impossible, such as accessing other people's calendars. This is fine, I just need to understand where we want to go.

From my PoV, I'm just trying not to add any more assumptions which we will regret later on. The `HasPrefix` checks are new here, and I don't think it's something we want to have long-term.

> Through https://github.com/emersion/go-webdav/pull/70 , I also realized that the backends should behave more like a HTTP backend (as opposed to e.g. map lookup or such).

Right. It's mostly about being able to return more detailed errors than "something failed". Just like `os.Open` errors can be introspected for instance.

> WDYT?

LGTM!

> I am quite disappointed that Go doesn't have a proper path.HasPrefix(), which is why I initially wrote those custom functions.

Yeah, I agree this is disappointing, as this is a common thing to do and it can be error-prone. Something like this can be used in backends:

```go
func pathHasPrefix(p, prefix string) bool {
    p = path.Clean(p)
    prefix = path.Clean(prefix)
    return p == prefix || strings.HasPrefix(p, prefix + "/")
}
```

Reply:
Removed the checks as discussed and also addressed your other comments.